test/cpp/util/spiffe_test_credentials.cc

```diff
@@ -0,0 +1,138 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "test/cpp/util/spiffe_test_credentials.h"
+#include "test/core/end2end/data/ssl_test_data.h"
+
+namespace grpc {
+namespace testing {
+
+class TestSyncTlsCredentialReload
+    : public ::grpc_impl::experimental::TlsCredentialReloadInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsCredentialReloadArg* arg) override {
+    struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair
+        pem_key_cert_pair = {test_server1_key, test_server1_cert};
+    std::vector<
+        struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair>
+        pem_key_cert_pair_list = {pem_key_cert_pair};
+    arg->set_key_materials(test_root_cert, pem_key_cert_pair_list);
+    arg->set_status(GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW);
+    return 0;
+  }
+};
+
+class TestSyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    arg->set_success(1);
+    arg->set_status(GRPC_STATUS_OK);
+    return 0;
+  }
+};
+
+static void TestAsyncTlsServerAuthorizationCheckCallback(
+    ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) {
+  GPR_ASSERT(arg != nullptr);
+  arg->set_success(1);
+  arg->set_status(GRPC_STATUS_OK);
+  arg->OnServerAuthorizationCheckDoneCallback();
+}
+
+class TestAsyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+ public:
+  TestAsyncTlsServerAuthorizationCheck(SpiffeThreadList* thread_list)
+      : thread_list_(thread_list) {}
+
+  // Async implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    server_authz_check_thread_ =
+        std::thread(TestAsyncTlsServerAuthorizationCheckCallback, arg);
+    thread_list_->add_thread(std::move(server_authz_check_thread_));
+    return 1;
+  }
+
+ private:
+  std::thread server_authz_check_thread_;
```

Comment:
Can we define it as a function local variable instead of class member?

Reply:
Yes, that's probably better, to just define it locally in the `Schedule` function. Good call.